src/pubsub.js

```diff
@@ -0,0 +1,569 @@
+              })
+            })
+          })
+
+          it('list with 3 subscribed topics', (done) => {
+            const topics = [{
+              name: 'one',
+              handler () {}
+            }, {
+              name: 'two',
+              handler () {}
+            }, {
+              name: 'three',
+              handler () {}
+            }]
+
+            each(topics, (t, cb) => {
+              ipfs1.pubsub.subscribe(t.name, t.handler, cb)
+            }, (err) => {
+              expect(err).to.not.exist
+              ipfs1.pubsub.ls((err, list) => {
+                expect(err).to.not.exist
+
+                expect(
+                  list.sort()
+                ).to.be.eql(
+                  topics.map((t) => t.name).sort()
+                )
+
+                topics.forEach((t) => {
+                  ipfs1.pubsub.unsubscribe(t.name, t.handler)
+                })
+
+                done()
+              })
+            })
+          })
+        })
+
+        describe('multiple nodes', () => {
+          it('receive messages from different node', (done) => {
+            const check = makeCheck(3, done)
+            const expectedString = 'hello from the other side'
+
+            const sub1 = (msg) => {
+              expect(msg.data.toString()).to.be.eql(expectedString)
+              // TODO: Reenable when go-ipfs is unbroken
+              // expect(msg.from).to.be.eql(ipfs2.peerId.id)
+              ipfs1.pubsub.unsubscribe(topic, sub1)
+              check()
+            }
+
+            const sub2 = (msg) => {
+              expect(msg.data.toString()).to.be.eql(expectedString)
+              // TODO: reenable when go-ipfs is unbroken
+              // expect(msg.from).to.be.eql(ipfs2.peerId.id)
+              ipfs2.pubsub.unsubscribe(topic, sub2)
+              check()
+            }
+
+            series([
+              (cb) => ipfs1.pubsub.subscribe(topic, sub1, cb),
+              (cb) => ipfs2.pubsub.subscribe(topic, sub2, cb),
+              (cb) => waitForPeers(ipfs2, topic, [ipfs1.peerId.id], cb)
+            ], (err) => {
+              expect(err).to.not.exist
+
+              ipfs2.pubsub.publish(topic, expectedString, check)
+            })
+          })
+
+          it('receive multiple messages', (done) => {
+            const inbox1 = []
+            const inbox2 = []
+            const outbox = ['hello', 'world', 'this', 'is', 'pubsub']
+
+            const check = makeCheck(outbox.length * 3, (err) => {
+              ipfs1.pubsub.unsubscribe(topic, sub1)
+              ipfs2.pubsub.unsubscribe(topic, sub2)
+
+              expect(inbox1.sort()).to.be.eql(outbox.sort())
+              expect(inbox2.sort()).to.be.eql(outbox.sort())
+
+              done(err)
+            })
+
+            function sub1 (msg) {
+              inbox1.push(msg.data.toString())
+              // TODO: enable when go-ipfs is unbroken
+              // expect(msg.from).to.be.eql(ipfs2.peerId.id)
```

Comment:
yes, it's mentioned in the first case, I just didn't want to duplicated it all over the file